src/datalakebundle/table/schema/SchemaChecker.py

```diff
@@ -1,51 +1,183 @@
+            raise Exception(error_message)
 
-            if "set_item_removed" in ddiff:
-                extra["missing_fields"] = self.__pretty_ordered_set_to_output(ddiff["set_item_removed"])
+    def check_metadata(self, table_definition: TableDefinition):
+        # self.__metadata_checker.check(table_definition)
+        pass
 
-            if "set_item_added" in ddiff:
-                extra["unexpected_fields"] = self.__pretty_ordered_set_to_output(ddiff["set_item_added"])
+    def generate_diff(self, df_schema: StructType, schema: StructType):
+        def remove_metadata(json_schema):
+            for field in json_schema["fields"]:
+                field["metadata"] = dict()
 
-            report = json.dumps(extra, indent=4).replace('"', "")
+            return json_schema
 
-            self.__logger.error(f"{error_message}\n{report}")
+        expected_schema = remove_metadata(schema.jsonValue())
+        df_schema = remove_metadata(df_schema.jsonValue())
 
-            raise Exception(error_message)
+        ddiff = DeepDiff(expected_schema, df_schema, ignore_string_case=True, ignore_order=True)
 
-    def __schema_to_set(self, schema: {}):
-        return set([(f["name"], f["type"]) for f in schema["fields"]])
+        result = []
+        if ddiff:
 
-    def __pretty_ordered_set_to_output(self, pos):
-        cols = [i[5:-1] for i in pos]
-        return cols
+            if "values_changed" in ddiff:
+                result.extend(self.__get_values_changed(expected_schema, ddiff["values_changed"]))
+
+            if "type_changes" in ddiff:
+                result.extend(self.__get_types_changed(expected_schema, ddiff["type_changes"]))
+
+            if "iterable_item_added" in ddiff:
+                result.extend(self.__get_iterable_item(expected_schema, ddiff["iterable_item_added"], "unexpected field"))
+
+            if "iterable_item_removed" in ddiff:
+                result.extend(self.__get_iterable_item(expected_schema, ddiff["iterable_item_removed"], "missing field"))
+
+            if "dictionary_item_added" in ddiff:
+                result.extend(self.__get_dictionary_item(expected_schema, ddiff["dictionary_item_added"], "Unexpected field"))
+
+            if "dictionary_item_removed" in ddiff:
+                result.extend(self.__get_dictionary_item(expected_schema, ddiff["dictionary_item_removed"], "Missing field"))
+
+        return result
+
+    def __rec_field_names(self, current: dict, chnks: list):
+        if chnks:
+            ch = chnks[0]
+            if len(ch) == 1 and len(chnks) == 1:
+                return
+            if ch[0] == "elementType":
+                yield "array"
+                yield from self.__rec_field_names(current["elementType"], chnks[1:])
+            else:
+                new = current[ch[0]][int(ch[1])]
+                yield new["name"]
+                yield from self.__rec_field_names(new["type"], chnks[1:])
```

Comment:
What is "rec" in the method name?

Reply:
The code in all the private methods bellow is very cryptic and does not have any tests.